### PR TITLE
[MLIR][TORCH] Add lowering for `aten._softmax` when `half_to_float=True`

### PR DESCRIPTION
**Pull request by Abhishek-Varma**
-- This commit adds decompose logic for `aten._softmax` when
   `half_to_float` is `True`.
-- I've added an MLIR lit test.
-- An e2e test case will be added once support for half to float conversion for
    `aten._softmax` is added.

Signed-off-by: Abhishek Varma <abhishek@nod-labs.com>

### CODE REVIEW

**Review by pashu123 — APPROVED** [review]
LGTM!

**Comment by silvasean** [comment]
Please remove the .mlir test and add an end-to-end test. See here for our policy: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

**Comment by pashu123** [comment]
> Please remove the .mlir test and add an end-to-end test. See here for our policy: https://github.com/llvm/torch-mlir/blob/main/docs/architecture.md#dos-and-donts-for-unit-vs-end-to-end-testing

@silvasean The e2e test for this op can't be added for the torch's cpu version and requires cuda version (float16 version of softmax). @Abhishek-Varma can add a test though, but it will be ultimately marked as xfail.

**Comment by silvasean** [comment]
Ah, got it. I would remove the .mlir test then and just mention that in the commit description.

**Comment by Abhishek-Varma** [comment]
Hi @silvasean - I've addressed your review comment.
1. Removed the .mlir test.
2. Updated the commit description to reflect the reason behind not adding an e2e test.

You may re-review/approve the PR.

File: lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp [hunk 1]
```diff
@@ -354,16 +353,26 @@ class DecomposeAten_SoftmaxOp : public OpRewritePattern<Aten_SoftmaxOp> {
...
-    // the same is not present on CPU.
-    if (halfToFloat)
-      return rewriter.notifyMatchFailure(
-          op, "halfToFloat is currently not supported.");
-
-    Value result = getSoftmaxResult(op, tensorType, rewriter);
+    BaseTensorType resultTensorType = op.getType().cast<BaseTensorType>();
+    // `torch.ops.aten._softmax`'s softmax with half to float conversion is not
+    // supported on CPU.
+    // But we go ahead with the decomposing and add a unit MLIR lit test.
```
Comment (silvasean) [inline lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp:359]: this comment seems out of date.

**Review by silvasean — APPROVED** [review]

**Pull request merged**